src/query/service/src/interpreters/interpreter_table_optimize.rs

```diff
@@ -45,81 +46,88 @@ impl Interpreter for OptimizeTableInterpreter {
-            .ctx
-            .get_table(&catalog_name, &db_name, &tbl_name)
-            .await?;
-
-        let action = self.plan.action.clone();
-        let do_purge = matches!(
-            action,
-            OptimizeTableAction::Purge(_) | OptimizeTableAction::All
-        );
-
-        let limit_opt = match action {
-            OptimizeTableAction::CompactBlocks(limit_opt) => limit_opt,
-            OptimizeTableAction::CompactSegments(limit_opt) => limit_opt,
-            _ => None,
-        };
+        let plan = self.plan.clone();
+        match self.plan.action.clone() {
+            OptimizeTableAction::CompactBlocks(limit_opt) => {
+                self.build_compact_pipeline(CompactTarget::Blocks, limit_opt)
+                    .await
+            }
+            OptimizeTableAction::CompactSegments(limit_opt) => {
+                self.build_compact_pipeline(CompactTarget::Segments, limit_opt)
+                    .await
+            }
+            OptimizeTableAction::Purge(point) => {
+                purge(ctx, plan, point).await?;
+                Ok(PipelineBuildResult::create())
+            }
+            OptimizeTableAction::All => {
+                let mut build_res = self
+                    .build_compact_pipeline(CompactTarget::Blocks, None)
+                    .await?;
 
-        let compact_target = match action {
-            OptimizeTableAction::CompactBlocks(_) | OptimizeTableAction::All => {
-                CompactTarget::Blocks
+                if build_res.main_pipeline.is_empty() {
+                    purge(ctx, plan, None).await?;
+                } else {
+                    build_res.main_pipeline.set_on_finished(move |may_error| {
+                        if may_error.is_none() {
+                            return GlobalIORuntime::instance()
+                                .block_on(async move { purge(ctx, plan, None).await });
+                        }
+                        Err(may_error.as_ref().unwrap().clone())
+                    });
+                }
+                Ok(build_res)
             }
-            OptimizeTableAction::CompactSegments(_) => CompactTarget::Segments,
-            _ => CompactTarget::None,
-        };
+        }
+    }
+}
 
+impl OptimizeTableInterpreter {
+    async fn build_compact_pipeline(
+        &self,
+        target: CompactTarget,
+        limit: Option<usize>,
+    ) -> Result<PipelineBuildResult> {
         let mut build_res = PipelineBuildResult::create();
+        let table = self
+            .ctx
+            .get_table(&self.plan.catalog, &self.plan.database, &self.plan.table)
+            .await?;
+        let table_info = table.get_table_info();
+        let catalog = self.ctx.get_catalog(&self.plan.catalog)?;
+        let res = catalog.get_table_mutation_lock(table_info).await?;
+        if res.locked {
+            return Err(ErrorCode::TableMutationAlreadyLocked(format!(
+                "table '{}' already locked for mutation",
```

Comment:
Consider change to 
table '{}' is under mutation, please retry compaction later
?